contracts/0.8.9/StakingRouter.sol

```diff
@@ -0,0 +1,610 @@
+        uint256 balance = address(this).balance;
+        if (balance > 0) {
+            getLido().receiveStakingRouter{value: balance}();
+        }
+    }
+
+    /**
+     * @notice Set credentials to withdraw ETH on Consensus Layer side after the phase 2 is launched to `_withdrawalCredentials`
+     * @dev Note that setWithdrawalCredentials discards all unused signing keys as the signatures are invalidated.
+     * @param _withdrawalCredentials withdrawal credentials field as defined in the Ethereum PoS consensus specs
+     */
+    function setWithdrawalCredentials(bytes32 _withdrawalCredentials) external onlyRole(MANAGE_WITHDRAWAL_CREDENTIALS_ROLE) {
+        WITHDRAWAL_CREDENTIALS_POSITION.setStorageBytes32(_withdrawalCredentials);
+
+        //trim keys with old WC
+        _trimUnusedKeys();
+
+        emit WithdrawalCredentialsSet(_withdrawalCredentials);
+    }
+
+    /**
+     * @notice Returns current credentials to withdraw ETH on Consensus Layer side after the phase 2 is launched
+     */
+    function getWithdrawalCredentials() public view returns (bytes32) {
+        return WITHDRAWAL_CREDENTIALS_POSITION.getStorageBytes32();
+    }
+
+    function _trimUnusedKeys() internal {
+        uint256 modulesCount = getStakingModulesCount();
+        for (uint256 i; i < modulesCount; ) {
+            IStakingModule(_getStakingModuleAddressByIndex(i)).invalidateReadyToDepositKeys();
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    function _loadStakingModuleCache(
+        uint256 _stakingModuleIndex
+    ) internal view returns (StakingModuleStatus status, StakingModuleCache memory stakingModuleCache) {
+        StakingModule storage stakingModuleData = _getStakingModuleByIndex(_stakingModuleIndex);
+        stakingModuleCache.stakingModuleAddress = stakingModuleData.stakingModuleAddress;
+        stakingModuleCache.moduleFee = stakingModuleData.moduleFee;
+        stakingModuleCache.treasuryFee = stakingModuleData.treasuryFee;
+        stakingModuleCache.targetShare = stakingModuleData.targetShare;
+        status = StakingModuleStatus(stakingModuleData.status);
+    }
+
+    /**
+     * @dev load not stopped modules list
+     * @notice used for reward distribution
+     * @return totalActiveKeys for not stopped modules
+     * @return modulesCache array of StakingModuleCache struct
+     */
+    function _loadNotStoppedStakingModulesCache()
+        internal
+        view
+        returns (uint256 totalActiveKeys, StakingModuleCache[] memory modulesCache)
+    {
+        uint256 modulesCount = getStakingModulesCount();
+        modulesCache = new StakingModuleCache[](modulesCount);
+        StakingModuleStatus status;
+        for (uint256 i; i < modulesCount; ) {
+            (status, modulesCache[i]) = _loadStakingModuleCache(i);
+
+            /// @dev account only keys from not stopped modules (i.e. active and paused)
+            if (status != StakingModuleStatus.Stopped) {
```

Comment:
Handle stopped module to allow flexible recover - let's send module's fee to the treasury

Reply:
_loadNotStoppedStakingModulesCache -> preprocessed + when setfee=0, add treasuryfee=treasury+modulefee